### PR TITLE
Feature/kvs health

### LINKED ISSUE #1819: Dropwizard healthchecks for Cassandra/database state
This is necessary so that the application is aware of the database state. We should have representation of the following 5 states:
1. The database is completely unreachable (No nodes are up)
2. The database is reachable but the configuration is out of sync with the database state. (eg: Replication factor is different)
3. Less than a quorum number of nodes are reachable. [No operations can be performed]
4. Quorum or more but not all are reachable. [Operations requiring quorum consistency will still run like reads/writes]
5. All nodes are reachable.

This healthcheck will be in the form of performing a simple operation like describe_keyspaces on each node. Also, each available node should have a valid partitioner.

This is a pre-requisite to #930.

### PR DESCRIPTION
**Pull request by hsaraogi**
**Goals (and why)**: Fixes #1819. Healthchecks for KVS status returned via the transactionManager.

**Implementation Description (bullets)**:
1. Implemented nodeAvailabilityStatus for cassandra. (All other KVSs are always available)
2. The transaction manager returns a state for the kvs based on the node availability.
3. ETE testing for Cassandra node availability.

TODO:
- [x] Implement terminal state.
- [x] nodeAvailability for Dbkvs
- [ ] env variables in tests have to be manually added again 

**Concerns (what feedback would you like?)**:
1. Is the overhead of trying to perform reads from Schema mutation lock table for healthcheck reasonable, we might get connection exceptions etc. There are absolutely no retries though.

**Where should we start reviewing?**:
1. SnapshotTransactionManager.getKeyValueServiceStatus()

**Priority (whenever / two weeks / yesterday)**: two weeks.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1832)

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 1]
```diff
@@ -2123,6 +2127,53 @@ public void compactInternally(TableReference tableRef) {
         }
     }
 
+    @Override
+    public NodeAvailabilityStatus getNodeAvailabilityStatus() {
+        try {
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: This method of getting availability status works in the standard configuration (3 nodes, RF3, or more generally numNodes = RF). However, if there are more nodes than the RF, then we're only checking whether we can do these reads for the `_locks` table - we could get lucky and return `EACH_QUORUM_AVAILABLE` when only 2 out of 6 nodes are up.

In the general case, it might be better to determine exactly how many nodes are up/down, using a similar healthcheck as is in CassandraClientPool.

Comment (hsaraogi) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: Yes, that makes more sense thinking about it again. I am planning to squash `EACH_QUORUM_AVAILABLE` and `LOCAL_QUORUM_AVAILABLE` to `QUORUM_AVAILABLE` and return quorum_available only when less than half RF nodes are down.

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/NodesDownTestSetup.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    static final Cell CELL_2_2 = Cell.create(SECOND_ROW, SECOND_COLUMN);
+    static final Cell CELL_3_1 = Cell.create(PtBytes.toBytes("row3"), FIRST_COLUMN);
+    static final Cell CELL_4_1 = Cell.create(PtBytes.toBytes("row4"), FIRST_COLUMN);
+
+    static final byte[] DEFAULT_CONTENTS = PtBytes.toBytes("default_value");
+    static final long DEFAULT_TIMESTAMP = 2L;
+    static final long OLD_TIMESTAMP = 1L;
+    static final Value DEFAULT_VALUE = Value.create(DEFAULT_CONTENTS, DEFAULT_TIMESTAMP);
+
+    static CassandraKeyValueService db;
```
Comment (hsaraogi) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/NodesDownTestSetup.java]: This was mostly a refactor, happy to rename.

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/NodesDownTestSetup.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        Awaitility.await()
+                .atMost(60, TimeUnit.SECONDS)
+                .until(NodesDownTestSetup::startupChecksPass);
+    }
+
+    private static boolean startupChecksPass() {
+        CassandraKeyValueServiceConfigManager manager = CassandraKeyValueServiceConfigManager.createSimpleManager(
+                ThreeNodeCassandraCluster.KVS_CONFIG);
+        try {
+            // startup checks are done implicitly in the constructor
```
Comment (hsaraogi) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/NodesDownTestSetup.java]: This is again a part of the refactor. Mostly making Test Setup common for one, two, three node down tests, we had only OneNodeDownTestSuite before. Not sure if the value will be seen in a separate PR.

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/NodesDownTestSetup.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        return true;
+    }
+
+    protected static void verifyValue(Cell cell, Value value) {
+        Map<Cell, Value> result = db.get(TEST_TABLE, ImmutableMap.of(cell, Long.MAX_VALUE));
+        assertThat(value).isEqualTo(result.get(cell));
+    }
+
+    protected static boolean tableExists(TableReference tableReference) {
+        Iterator<TableReference> it = NodesDownTestSetup.db.getAllTableNames().iterator();
```
Comment (hsaraogi) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/NodesDownTestSetup.java]: Again, was a refactor but sure this is much cleaner :).

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/AbstractNodeAvailabilityTest.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+package com.palantir.cassandra.multinode;
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.NodeAvailabilityStatus;
+
+public abstract class AbstractNodeAvailabilityTest {
```
Comment (hsaraogi) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/AbstractNodeAvailabilityTest.java]: A parametrized class will run with, say, 3 sets of parameters under the same setup. This means one node down will run `NO_QUORUM_AVAILABLE` and `LOCAL_QUORUM_AVAILABLE`. This is not the intent here we want one node down to run `LOCAL_QUORUM_AVAILABLE` and 2/3 nodes down to only run `NO_QUORUM_AVAILABLE`.

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/NodesDownTestSetup.java (hunk 4 above)
Comment (gsheasby) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/NodesDownTestSetup.java]: Yep, ack'd that this code was pulled out from a different place. Thanks for cleaning it up! :)

File: atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/AbstractNodeAvailabilityTest.java (hunk 5 above)
Comment (gsheasby) [inline atlasdb-cassandra-multinode-tests/src/test/java/com/palantir/cassandra/multinode/AbstractNodeAvailabilityTest.java]: Makes sense - I see that the impls need to be set up in different suites so the KVS is set up correctly.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 6]
```diff
@@ -2129,53 +2126,60 @@ public void compactInternally(TableReference tableRef) {
 
     @Override
     public NodeAvailabilityStatus getNodeAvailabilityStatus() {
-        Map<InetSocketAddress, CassandraClientPoolingContainer> currentPools = clientPool.currentPools;
-        int replicationFactor = configManager.getConfig().replicationFactor();
+        if (!doesConfigReplicationFactorMatchWithCluster()) {
+            return NodeAvailabilityStatus.TERMINAL;
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: We should consider what we expect a client to do when they get NodeAvailiabilityStatus.TERMINAL - especially as there are a couple of distinct reasons why we give that phase, and both require specific config changes. Do we want to pass a message along with the `TERMINAL` status?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 7]
```diff
@@ -2129,53 +2126,60 @@ public void compactInternally(TableReference tableRef) {
...
+        if (numberOfUnreachableNodes == 0) {
+            return NodeAvailabilityStatus.ALL_AVAILABLE;
+        } else if (fewerThanRfByTwoNodesUnreachable(numberOfUnreachableNodes)) {
+            return NodeAvailabilityStatus.QUORUM_AVAILABLE;
+        } else {
+            return NodeAvailabilityStatus.NO_QUORUM_AVAILABLE;
+        }
+    }
+
+    private boolean fewerThanRfByTwoNodesUnreachable(int numberOfUnreachableNodes) {
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: This method name describes perfectly _what_ the method does, but not _why_. The "what" can be readily understood by reading the code, so how about `isQuorumAvailable`?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraVerifier.java [hunk 8]
```diff
@@ -286,6 +286,17 @@ private static void checkAndSetReplicationFactor(
...
+    }
+
+    static void currentRfOnKeyspaceMatchesDesiredRf(Client client, CassandraKeyValueServiceConfig config,
+            boolean safetyDisabled) throws TException {
+        KsDef ks = client.describe_keyspace(config.keyspace());
+        Set<String> dcs = sanityCheckDatacenters(client, config.replicationFactor(), safetyDisabled);
+        sanitiseReplicationFactor(ks, config.replicationFactor(), safetyDisabled, dcs);
+    }
+
+    private static void sanitiseReplicationFactor(KsDef ks, int desiredRf, boolean safetyDisabled, Set<String> dcs) {
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraVerifier.java]: `sanityCheckReplicationFactor`?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 9]
```diff
@@ -2129,53 +2126,60 @@ public void compactInternally(TableReference tableRef) {
...
-                    lockRowName,
-                    lockColName);
-            return runCqlQuery(selectCql, client, consistency);
+    private boolean doesConfigReplicationFactorMatchWithCluster() {
+        return clientPool.run(client -> {
+            try {
+                CassandraVerifier.currentRfOnKeyspaceMatchesDesiredRf(client, configManager.getConfig(), false);
+                return true;
+            } catch (Exception e) {
+                return false;
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: For sure, you should log the exception message instead of just swallowing it.

File: docs/source/release_notes/release-notes.rst [hunk 10]
```diff
@@ -42,6 +42,12 @@ develop
     *    - Type
          - Change
 
+
+    *    - |improved|
+         - ``TransactionManager`` now has an API ``getKeyValueServiceStatus`` that can return the health of the underlying KVS. This is designed for applications
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: Should add documentation. What does each status mean, and what is the application developer expected to do to handle each case?

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionManager.java [hunk 11]
```diff
@@ -106,6 +106,13 @@
     long getImmutableTimestamp();
 
     /**
+     * Return true if KVS is healthy, else return false.
```
Comment (gsheasby) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionManager.java]: Inaccurate - doesn't just return true/false any more.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java (hunk 6 above)
Comment (hsaraogi) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: `TERMINAL` state implies the system is in a state that cannot be recovered without manual intervention and the box should report shutdown or something. The logs will have the info as to what is the reason behind `TERMINAL`.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 12]
```diff
@@ -2123,6 +2124,65 @@ public void compactInternally(TableReference tableRef) {
         }
     }
 
+    @Override
+    public NodeAvailabilityStatus getNodeAvailabilityStatus() {
+        if (!doesConfigReplicationFactorMatchWithCluster()) {
+            return NodeAvailabilityStatus.TERMINAL;
+        }
+        return getStatusByRunningOperationsOnEachHost();
```
Comment (tpetracca) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: I think the ordering is backwards here. You're checking that keyspace config is correct before you're even sure that the cluster is up. What happens if all nodes are down? Will `doesConfigReplicationFactorMatchWithCluster()` return true because it can't tell or false because the operation throws an exception because it can't reach the node.

Comment (tpetracca) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: My guess is we're best off either reversing the order here or doing all the actual terminal checks after we've confirmed a node is healthy like how the partitioner check is currently being ran.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 13]
```diff
@@ -2123,6 +2124,65 @@ public void compactInternally(TableReference tableRef) {
         }
     }
 
+    @Override
+    public NodeAvailabilityStatus getNodeAvailabilityStatus() {
+        if (!doesConfigReplicationFactorMatchWithCluster()) {
```
Comment (tpetracca) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: Ah I got further down and saw that this indeed does all the actual checks and takes `safetyDisabled` into account.
